### PR TITLE
Extract NodePattern to a separated gem

### PR DESCRIPTION
**Pull request by jonatas**
I started the work related to #6655.

The gem extracted is under my namespace and I can move it soon some maintainer transfer it.

https://github.com/jonatas/rubocop-node_pattern

The name of the gem is also in the TODO  list to be discussed and probably we can find a better name since all AST utilities are going to the repo too.

## TODO

* [ ] Move the related project to `rubocop-hq` namespace
* [ ] Update the Gemfile to `rubocop-hq` namespace 
* [ ] Move NodePattern implementation to `NodePattern::Parser` or something to make the gem clear and avoid [requires in the bottom](https://github.com/jonatas/rubocop-node_pattern/blob/master/lib/rubocop/node_pattern.rb#L567)
* [ ] Define a better name to the gem?
* [ ] Setup RuboCop config for the project

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

## Notes

I'm not sure how can we maintain the ownership of the current git files that are moving to another repo. Is it something common or we just move and have a new author?

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -14,6 +14,9 @@ gem 'simplecov', '~> 0.10'
 gem 'test-queue'
 gem 'yard', '~> 0.9'
 
+gem 'rubocop-node_pattern',
+    git: 'git://github.com/jonatas/rubocop-node_pattern.git'
```
Comment (jonatas) [inline Gemfile]: Soon we migrate the gem to the proper repo, we can also publish as a gem and remove the git reference.

Comment (wevtimoteo) [reply Gemfile]: Did you consider keeping git history using filter-branch feature for those files?

https://help.github.com/articles/splitting-a-subfolder-out-into-a-new-repository/

Comment (jonatas) [reply Gemfile]: That  is pretty cool @wevtimoteo, I'll use the tool to keep the git history intact 👍

**Comment by bbatsov** [comment]
Keep in mind that the node extensions and the note pattern matcher are somewhat orthogonal - you don’t necessary always need the two of them. This means that potentially we can have an `astx` gem and some other gem about the pattern matcher. On the other hand this would mean more overhead from a maintenance perspective, so perhaps a single gem would do. I also think it might be convenient to simply have those 2 or 3 gems in the same repo (RuboCop’s) for simplicity’s sake. We will have to think this over very carefully.

**Comment by jonatas** [comment]
Got it @bbatsov. Looking at the code, we can see `node_pattern` depends on `astx`. We can do it in two gems or only one.

@baweaver let me know how can we schedule a pairing session or work together on it. I'm open to both cases. We can go with one or two gems. I'd start with one and we can split the external gem later if needed.

File: Gemfile (hunk 1 above)
Comment (wevtimoteo) [reply Gemfile]: If you need some help with that I'll happy to pair with you!

**Comment by pirj** [comment]
With the movement of extracting parts of RuboCop out of it (e.g. `rubocop-performance`), while keeping those extracted repositories first-class citizen of RuboCopHQ, it makes sense to think about how a change to main `rubocop`, and specifically to its node pattern repository would affect the dependent repositories.

An approach to consider is [the one taken by RSpec]( https://github.com/rspec/rspec-dev#rspec-dev), when all the repositories are checked out, and before submitting a change one would check if the dependent officially supported gems are not broken.
```
# rspec-expectations' Gemfile
branch = File.read(File.expand_path("../maintenance-branch", __FILE__)).chomp # `master`, or some cross-repository branch
%w[rspec rspec-core rspec-mocks rspec-support].each do |lib|
  library_path = File.expand_path("../../#{lib}", __FILE__)
  if File.exist?(library_path) && !ENV['USE_GIT_REPOS']
    gem lib, :path => library_path
  else
    gem lib, :git => "https://github.com/rspec/#{lib}.git", :branch => branch
  end
end
```

With that modular yet atomically safe approach, it doesn't matter that much anymore if `rubocop-node_pattern` is in the same repository or not, and breakages can be spotted early, during local development. That also kind of simplifies the changes that need to be made in several repositories at once to fix a bug/introduce a feature.

WDYT?

**Comment by bbatsov** [comment]
Yeah, that's a reasonable approach. We should first finish the isolation/extraction of something though. :-) Right now everything's work in progress...

@jonas054 Any updates here?

**Comment by jonas054** [comment]
@bbatsov It's very nice to hear from you again! :smile: But did you in this case mean @jonatas?

**Comment by bbatsov** [comment]
Ops! Sorry about that! I totally meant to ping @jonatas, but I love you so much that I type your name every time I start typing `@jona...`. 😄

**Comment by jonatas** [comment]
I like the idea and I guess we can try to follow the approach @pirj suggested.

The hardest part is that we have `ast/node` that is a dependency for `node_pattern`. I'll try to pair with @pirj tomorrow and see if we can handle it.

**Comment by bbatsov** [comment]
Sounds good. Ping me if you need my help with anyone.

**Comment by jonatas** [comment]
Thanks, @bbatsov. Should we keep the same module name for now or already go with `astx`?

File: Gemfile (hunk 1 above)
Comment (wevtimoteo) [reply Gemfile]: Done in https://github.com/jonatas/rubocop-astx/tree/move-ast-keeping-history

**Comment by bbatsov** [comment]
I guess for now it can be the same module name.

**Comment by marcandre** [comment]
Let me know if you need help...

**Comment by jonatas** [comment]
I'm not having the proper time to finish it. I'll close it for now and revisit in a near future.

**Comment by jonatas** [comment]
@marcandre feel free to ping me I can try to pair with you and help to review it.

**Pull request closed without merging**